File: ios/NfWifiTask.m

#import "NfWifiTask.h"

#import "NfWifiSocketServer.h"
#import "NfWifiSocketClient.h"
#import "NfWifiGenerator.h"

@interface NfWifiTask()

@property(nonatomic, strong) NSString* _ssid;
@property(nonatomic, strong) NSString* _key;
@property(nonatomic, strong) NfWifiSocketClient* _client;
@property(nonatomic, strong) NfWifiSocketServer* _server;
@property(nonatomic, assign) volatile BOOL _isClosed;
@property(nonatomic, copy) void (^_callback)(NSString* evt);

@end

@implementation NfWifiTask

-(id) initWithSsid: (NSString*) ssid withKey:(NSString*) key {
    
    self = [super init];
    
    if(self){
        self._ssid = ssid;
        self._key = key;
        self._client = [[NfWifiSocketClient alloc] init];
        self._server = [[NfWifiSocketServer alloc] initWithPort:60002];
    }
    
    return self;
}

-(void)start:(void(^)(NSString* evt))callback;
{
  self._callback = callback;
  self._isClosed = false;
  dispatch_queue_t queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
  dispatch_async(queue, ^{
    //Byte receiveOneByte = -1;
    Byte receiveAll[24];
    while(!self._isClosed){
      NSData* receiveData = [self._server receiveSpecLenBytes:24];
      
      if(receiveData){
        [receiveData getBytes:receiveAll length: 24];
        
        if(receiveAll[0] == 0x18
           && receiveAll[8] == 0x06 && receiveAll[9] == 0x00
           && receiveAll[10] == 0x01){
          
          NSString * deviceId = [NSString stringWithFormat:@"%02x%02x%02x%02x%02x%02x", receiveAll[11], receiveAll[12], receiveAll[13], receiveAll[14], receiveAll[15], receiveAll[16] ];
          self._callback(deviceId);
        }
      }else
      {
        break;
      }
    }
  });
  
  dispatch_async(queue, ^{
    NfWifiGenerator* generator = [[NfWifiGenerator alloc]initWithSsid:self._ssid withKey:self._key];
    NSArray* bytesArray = [generator getBytesArray];
    while(!self._isClosed){
      [self._client sendDataWithBytesArray:bytesArray withTarget:@"239.1.2.110" withPort:60001 withInterval:1];
    }
    
  });
}

- (void) stop
{
  self._isClosed = true;
  if(self._server){
    [self._server stop];
  }
  
  if(self._client){
    [self._client stop];
  }
}

@end

File: android/src/main/java/com/lumin824/nufrontwifi/NfWifiTask.java
package com.lumin824.nufrontwifi;

import android.os.Handler;
import android.os.Message;
import android.util.Log;

import java.util.List;

/**
 * Created by lumin on 16/5/29.
 */
public class NfWifiTask {

    private String ssid;
    private String key;
    private NfWifiSocketClient client;
    private NfWifiSocketServer server;

    private Thread clientThread;
    private Thread serverThread;

    private boolean isRunning;

    public NfWifiTask(String ssid, String key) {
        this.ssid = ssid;
        this.key = key;
        this.client = new NfWifiSocketClient();
        this.server = new NfWifiSocketServer(60002);
    }

    public void start(final Handler callback){

        isRunning = true;
        serverThread = new Thread(new Runnable(){
            @Override
            public void run() {
            while(isRunning){
                byte[] receiveAll = server.receiveSpecLenBytes(24);
                if(receiveAll != null && receiveAll[0] == 0x18
                    && receiveAll[8] == 0x06 && receiveAll[9] == 0x00
                    && receiveAll[10] == 0x01){
                    String deviceId = String.format("%02x%02x%02x%02x%02x%02x", receiveAll[11], receiveAll[12], receiveAll[13],
                        receiveAll[14], receiveAll[15], receiveAll[16]);
                    Message msg = callback.obtainMessage();
                    msg.obj = deviceId;
                    callback.sendMessage(msg);
                }
            }
            Log.d("rnwifi", "server thread is stop");
            }
        });
        serverThread.start();

        clientThread = new Thread(new Runnable() {
            @Override
            public void run() {
            NfWifiGenerator generator = new NfWifiGenerator(ssid, key);
            List<byte[]> bytesArray = generator.getBytesArray();
            String target = "239.1.2.110";
            int port = 60001;
            int interval = 1;
            while(isRunning){
                client.sendDataWithBytesArray(bytesArray, target, port, interval);
            }
            Log.d("rnwifi", "client thread is stop");
            }
        });
        clientThread.start();
    }

    public void stop(){
        Log.d("rnwifi", "stop");
        this.isRunning = false;
        if(server != null) server.close();
        if(client != null) client.close();
    }
}
